### PR TITLE
Converted mixBerMnistEM to python

### LINKED ISSUE #174: Convert mixBerMnistEM to python
This is figure number 21.19 in the book. its matlab code is [here](https://github.com/probml/pmtk3/blob/master/demos/mixBerMnistEM.m)

### PR DESCRIPTION
**Pull request by shivaditya-meduri**
mixBernoulliMnist

Closes #174

The code might take some time to load and this is because it is downloading the MNIST dataset from online and the delay in execution is solely due to this and not due to the inefficiency of the code.

The values and the images will differ because the data is randomly chosen from the total set of images.

I initially added +1 to all images in the training set and changed the Bernoulli function to adapt to this, but did not see any change in the execution, It was more convenient for me to use the binary image as is and use the implementation of Bernoulli distribution which is used for this mixture model.

In this implementation, I did not find the optimal k because the optimal k=20 has already been found through trial and error in the MATLAB code and no graph implementation is found in the book.

Implemented the changes suggested:
1. Vectorized code and replaced iterating over the data points to vectorized versions
2. replaced Bernoulli function with https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bernoulli.html
3. factored out the code that does parameter initialization and image plotting into subfunctions, to simplify main()
4. changed the data source from open_ml to tensorflow.datasets

### CODE REVIEW

**Comment by shivaditya-meduri** [comment]
Hi @murphyk , @mjsML , I made the suggested changes. Please review this PR

**Review by murphyk — CHANGES_REQUESTED** [review]
- The demo is very slow, because full batch EM on 60k examples is a bad idea. In the matlab code, I randomly subsampled N=1000 examples (after shuffling the rows); please do the same,
- Please expand the name 'resp' to 'responsibilities', and use PEP8 naming, not camel case, so `respBernoulli` becomes `compute_responsibilitie`, `bernoulliMStep` becomes `m_step`, `llBernoulli` becomes `expected_log_lik`, etc.
- Rename `bernoulli` (which is ambiguous) to `bernoulli_mixture_pmf`.

**Comment by shivaditya-meduri** [comment]
Hi @murphyk, Did you mean that the current python code attached with this PR is very slow and I need to look into it or it is sufficiently fast enough and I could change the N to 10k?

**Comment by murphyk** [comment]
It should suffice to just set N=10k. (The code seems as  vectorized as it can be, without resorting to jax...)

**Comment by shivaditya-meduri** [comment]
Thank you for your time reviewing this code, I will make the changes ASAP

**Comment by shivaditya-meduri** [comment]
Hi @murphyk , I changed N to 10k and the execution time increased almost 8 fold, Initially it was around 2 mins(N=1000), now it takes around 15 mins(N=10k) for execution. I am currently trying JAX to implement a faster version of the mixture of Bernoulli model.
What would be the limit before the execution is considered too slow? Depending on that we can train the model to the said number of images.

**Comment by murphyk** [comment]
I like all the demos to run in under ~2 mins :). Using N=1000 and max_iter  = 5 meets that goal :). As for JAX, this may help, but the biggest speedup will come from an online EM algorithm, rather than full batch. For this I would suggest using stochastic variational inference. I will cover this in vol 2, and make an issue :)

**Comment by shivaditya-meduri** [comment]
Thank you for your pateince!

**Comment by murphyk** [comment]
If you like to get a head-start, a JAX version of this SVI code (that uses TFP) would be great: https://brendanhasz.github.io/2019/06/12/tfp-gmm.html

**Comment by shivaditya-meduri** [comment]
Thanks @murphyk, I will start working on it.

**Comment by murphyk** [comment]
I opened https://github.com/probml/pyprobml/issues/348 to track progress on this.
